tempesta_fw/msg.c

```diff
@@ -34,20 +34,22 @@
 int
 tfw_msg_write(TfwMsgIter *it, const TfwStr *data)
 {
-	char *p;
 	const TfwStr *c, *end;
-	skb_frag_t *frag;
-	unsigned int c_off = 0, f_size, c_size, f_room, n_copy;
 
 	BUG_ON(TFW_STR_DUP(data));
+	if (WARN_ON_ONCE(it->frag >= skb_shinfo(it->skb)->nr_frags))
+		return -E2BIG;
+
 	TFW_STR_FOR_EACH_CHUNK(c, data, end) {
+		char *p;
+		unsigned int c_off = 0, c_size, f_room, n_copy;
 this_chunk:
-		if (!frag)
-			return -E2BIG;
 
 		c_size = c->len - c_off;
 		if (it->frag >= 0) {
-			frag = &skb_shinfo(it->skb)->frags[it->frag];
+			unsigned int f_size;
+			skb_frag_t *frag = &skb_shinfo(it->skb)->frags[it->frag];
+
```

Comment:
It seems that there is a problem with using skb header when allocating more than one skb - in `tfw_http_msg_setup()` each skb is allocated with a place under the header:
https://github.com/tempesta-tech/tempesta/blob/5631d72b222aa62fdd74bc5fe56650d687a04042/tempesta_fw/ss_skb.h#L174
but in `tfw_msg_write()` only the first skb header in the list is used (`it->frag < 0`):
https://github.com/tempesta-tech/tempesta/blob/5631d72b222aa62fdd74bc5fe56650d687a04042/tempesta_fw/msg.c#L59
for subsequent skb `it-> frag` is always initialized by zero:
https://github.com/tempesta-tech/tempesta/blob/5631d72b222aa62fdd74bc5fe56650d687a04042/tempesta_fw/ss_skb.h#L132
so then we always fall into the `it->frag >= 0` branch.
Look like the current master version of `tfw_msg_write()` has the same issue too.